File: android/app/src/main/java/com/lacour/vincent/wificaresp8266/connector/CarConnector.kt
package com.lacour.vincent.wificaresp8266.connector

import android.content.Context
import android.util.Log
import android.widget.Toast
import com.lacour.vincent.wificaresp8266.R
import com.lacour.vincent.wificaresp8266.storage.Preferences
import retrofit2.*

import retrofit2.http.GET
import retrofit2.http.Query


private interface CarApiService {
    @GET("/move")
    fun move(@Query("dir") direction: String): Call<Void>

    @GET("/action")
    fun action(@Query("type") type: String): Call<Void>
}

class CarConnector(context: Context) {

    private val preferences = Preferences(context)
    private var service: CarApiService? = null

    init {
        val url = "http://${preferences.getIpAddress()}:${preferences.getPort()}"
        try {
            val retrofit = Retrofit.Builder().baseUrl(url).build()
            service = retrofit.create(CarApiService::class.java)
        } catch (e: IllegalArgumentException) {
            Toast.makeText(
                context,
                context.getString(R.string.invalid_hostname, url),
                Toast.LENGTH_LONG
            ).show()
        }

    }

    fun moveForward() = sendMoveRequest(preferences.getMoveForwardValue())
    fun moveBackward() = sendMoveRequest(preferences.getMoveBackwardValue())
    fun stopMoving() = sendMoveRequest(preferences.getStopValue())
    fun turnLeft() = sendMoveRequest(preferences.getTurnLeftValue())
    fun turnRight() = sendMoveRequest(preferences.getTurnRightValue())

    fun actionOne() = sendActionRequest(preferences.getActionOneValue())
    fun actionTwo() = sendActionRequest(preferences.getActionTwoValue())
    fun actionThree() = sendActionRequest(preferences.getActionThreeValue())
    fun actionFour() = sendActionRequest(preferences.getActionFourValue())
    fun actionFive() = sendActionRequest(preferences.getActionFiveValue())
    fun actionSix() = sendActionRequest(preferences.getActionSixValue())
    fun actionSeven() = sendActionRequest(preferences.getActionSevenValue())
    fun actionHeight() = sendActionRequest(preferences.getActionHeightValue())


    private fun sendMoveRequest(dir: String) {
        if (service == null) return
        val request = (service as CarApiService).move(dir)
        request.enqueue(object : Callback<Void> {
            override fun onResponse(call: Call<Void>, response: Response<Void>) {
                Log.i("Response", response.code().toString())
            }

            override fun onFailure(call: Call<Void>, t: Throwable) {
                Log.e("Failure", t.message.toString())
            }
        })
    }

    private fun sendActionRequest(type: String) {
        if (service == null) return
        val request = (service as CarApiService).action(type)
        request.enqueue(object : Callback<Void> {
            override fun onResponse(call: Call<Void>, response: Response<Void>) {
                Log.i("Response", response.code().toString())
            }

            override fun onFailure(call: Call<Void>, t: Throwable) {
                Log.e("Failure", t.message.toString())
            }
        })
    }

}

File: ios/wificaresp8266/Connector/CarConnector.swift
import Alamofire

struct CarConnector {
    let preferences = Preferences()
    var apiUrl: String

    init() {
        apiUrl = "http://\(preferences.getIpAddress()):\(preferences.getPort())"
    }

    func moveForward() { sendMoveRequest(dir: preferences.getMoveForwardValue()) }
    func moveBackward() { sendMoveRequest(dir: preferences.getMoveBackwardValue()) }
    func stopMoving() { sendMoveRequest(dir: preferences.getStopValue()) }
    func turnRight() { sendMoveRequest(dir: preferences.getTurnRightValue()) }
    func turnLeft() { sendMoveRequest(dir: preferences.getTurnLeftValue()) }

    func actionOne() { sendActionRequest(type: preferences.getActionOneValue()) }
    func actionTwo() { sendActionRequest(type: preferences.getActionTwoValue()) }
    func actionThree() { sendActionRequest(type: preferences.getActionThreeValue()) }
    func actionFour() { sendActionRequest(type: preferences.getActionFourValue()) }
    func actionFive() { sendActionRequest(type: preferences.getActionFiveValue()) }
    func actionSix() { sendActionRequest(type: preferences.getActionSixValue()) }
    func actionSeven() { sendActionRequest(type: preferences.getActionSevenValue()) }
    func actionHeight() { sendActionRequest(type: preferences.getActionHeightValue()) }

    private func sendMoveRequest(dir: String) {
        let url = "\(apiUrl)/move?dir=\(dir)"
        AF.request(url, method: .get)
            .responseString { response in
                print(response)
            }
    }

    private func sendActionRequest(type: String) {
        let url = "\(apiUrl)/action?type=\(type)"
        AF.request(url, method: .get)
            .responseString { response in
                print(response)
            }
    }
}
